File: hackerrank/data_structures/heap/median.cc
#include <bits/stdc++.h>

using namespace std;

/*
 * Complete the runningMedian function below.
 * https://www.hackerrank.com/challenges/find-the-running-median/forum/comments/456990
 */
vector<double> runningMedian(vector<int> a)
{
  vector<double> res;
  priority_queue<int, vector<int>, greater<int>> minheap;
  priority_queue<int> maxheap;

  double median = 0;

  for (int i = 0; i < a.size(); i++)
  {
    if (a[i] <= median)
    {
      maxheap.push(a[i]);
    }
    else
    {
      minheap.push(a[i]);
    }

    if (minheap.size() > maxheap.size() + 1)
    {
      maxheap.push(minheap.top());
      minheap.pop();
    }
    if (maxheap.size() > minheap.size() + 1)
    {
      minheap.push(maxheap.top());
      maxheap.pop();
    }

    if (minheap.size() == maxheap.size())
    {
      median = (maxheap.top() + minheap.top()) / 2.0;
    }
    else if (minheap.size() > maxheap.size())
    {
      median = (double)minheap.top();
    }
    else if (minheap.size() < maxheap.size())
    {
      median = (double)maxheap.top();
    }
    res.push_back(median);
  }
  return res;
}
int main()
{
  ofstream fout(getenv("OUTPUT_PATH"));

  int a_count;
  cin >> a_count;
  cin.ignore(numeric_limits<streamsize>::max(), '\n');

  vector<int> a(a_count);

  for (int a_itr = 0; a_itr < a_count; a_itr++)
  {
    int a_item;
    cin >> a_item;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    a[a_itr] = a_item;
  }

  vector<double> result = runningMedian(a);

  for (int result_itr = 0; result_itr < result.size(); result_itr++)
  {
    fout << result[result_itr];

    if (result_itr != result.size() - 1)
    {
      fout << "\n";
    }
  }

  fout << "\n";

  fout.close();

  return 0;
}

File: hackerrank/data_structures/heap/median.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

/* https://www.hackerrank.com/challenges/find-the-running-median/problem
 * Complete the runningMedian function below.
 * TODO: this could be solved following https://coder.today/tech/2017-10-13_fast-priority-queues-in-golang-hierarchical-heap/
 */

type heap struct {
	Value    int32
	Size     int32
	Previous *heap
	Next     *heap
}

func (h *heap) swap(j *heap) {
	tmp := h.Previous
	h.Next = j
	j.Previous = tmp
	j.Next = h
}

func emptyHeap() *heap {
	return &heap{
		Size: 0,
		Next: nil,
	}
}

func (h *heap) add(val int32) *heap {
	return &heap{
		Value: val,
		Size:  h.Size + 1,
		Next:  h,
	}
}

func median(min *heap, max *heap) float64 {

}

func runningMedian(a []int32) []float64 {
	/*
	 * Write your code here.
	 */
	var median float64
	minHeap := &heap{
		Size: 0,
		Next: nil,
	}
	maxHeap := &heap{}

}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 1024*1024)

	aCount, err := strconv.ParseInt(readLine(reader), 10, 64)
	checkError(err)

	var a []int32

	for aItr := 0; aItr < int(aCount); aItr++ {
		aItemTemp, err := strconv.ParseInt(readLine(reader), 10, 64)
		checkError(err)
		aItem := int32(aItemTemp)
		a = append(a, aItem)
	}

	result := runningMedian(a)

	for resultItr, resultItem := range result {
		fmt.Fprintf(writer, "%f", resultItem)

		if resultItr != len(result)-1 {
			fmt.Fprintf(writer, "\n")
		}
	}

	fmt.Fprintf(writer, "\n")

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
